content/en/blog/_posts/2023-12-07-a-demo-of-webassembly-and-wasi-on-kubernetes/index.md

```diff
@@ -0,0 +1,880 @@
+
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: wasm-demo-app
+  annotations:
+    module.wasm.image/variant: compat
+spec:
+  runtimeClassName: crun
+  containers:
+  - name: wasm-demo-app
+    image: docker.io/cr7258/wasm-demo-app:v1
+```
+
+You can use `kubectl port-forward` to forward traffic from your local machine into the Kubernetes cluster:
+
+```bash
+kubectl port-forward pod/wasm-demo-app 8080:8080
+```
+
+Open a new terminal, send a request to the service.
+
+```bash
+curl http://localhost:8080
+```
+
+The output is:
+
+```bash
+Hello, World!
+```
+
+Once the testing is complete, you can destroy the cluster by running:
+
+```bash
+kind delete cluster --name wasm-demo
+```
+##### Set up your cluster for Wasm automatically using KWasm, then run the app inside a pod
+
+_[Kwasm](https://kwasm.sh/)_ is a Kubernetes Operator that adds WebAssembly support to your Kubernetes nodes. The operator uses the _[kwasm-node-installer](https://github.com/KWasm/kwasm-node-installer)_ project to modify the underlying Kubernetes nodes.
+
+In this section, I will demostrate how to use Kwasm Operator to add Wasm support to Kubernetes nodes automatically. To enable Wasm support on a particular node, simply add the annotation `kwasm.sh/kwasm-node=true` on that node. This will trigger Kwasm to create a Job to deploy the necessary binary files needed to run Wasm on the node. Additionally, containerd's configuration will be modified accordingly.
+
+{{< figure src="05-kwasm-operator.svg" alt="The diagram illustrates that when Kwasm Operator detects the annotation kwasm.sh/kwasm-node=true on a Kubernetes node, it will create a Job to enable Wasm support on that node." caption="Figure 5: Kwasm Operator" >}}
+
+Create a single-node Kubernetes cluster using kind.
+
+```bash
+kind create cluster --name kwasm-demo
+```
+
+A Helm chart is available to easily install the Kwasm operator. Install the Kwasm Operator using helm and enable Wasm support for the node `kwasm-demo-control-plane` by adding the annotation `kwasm.sh/kwasm-node=true`.
+
+```bash
+# Add Helm repository if not already done
+helm repo add kwasm http://kwasm.sh/kwasm-operator/
+# Install KWasm operator
+helm install -n kwasm --create-namespace kwasm-operator kwasm/kwasm-operator
+# Provision Nodes
+kubectl annotate node kwasm-demo-control-plane kwasm.sh/kwasm-node=true
+```
+
+Add label `runtime=crun-kwasm` on the node.
+
+```bash
+kubectl label nodes kwasm-demo-control-plane runtime=crun-kwasm
+```
+
+Create a RuntimeClass resource named `crun` to use the crun handler automatically set up by Kwasm in containerd, the `scheduling.nodeSelector` property sends pod to nodes with the `runtime=crun-kwasm` label. Here is an example manifest:
+
+```yaml
+apiVersion: node.k8s.io/v1
+kind: RuntimeClass
+metadata:
+  name: crun
+scheduling:
+  nodeSelector:
+    runtime: crun-kwasm
+handler: crun
+```
```

Comment:
@cr7258, @sftim, IMPORTANT NOTE:
KWasm by default uses the latest [https://github.com/KWasm/kwasm-node-installer](kwasm-node-installer) image built from the `main' branch. Which at the time of writing is > 0.3.0 

> [!CAUTION]
> KWasm Node Installer version v0.3.0 has removed crun in favor of the WasmEdge shim.

There are now two options:

1. Use the v0.2.0 version of the installer.
2. Change the RuntimeClass to `wasmedge`.

## Using the v0.2.0 installer image
```
helm upgrade --install -n kwasm --create-namespace kwasm-operator kwasm/kwasm-operator \
  --set kwasmOperator.autoProvision="true" \
  --set kwasmOperator.installerImage="ghcr.io/kwasm/kwasm-node-installer:v0.2.0"
```

## Changing the RuntimeClass
The `wasmedge` shim behaves like `crun+wasmedge`, but it doesn't need the `compat/compat-smart` annotation. It always uses the Linux container as a fallback. All images that work with `crun+wasmedge` should also work with the shim.

```yaml
apiVersion: node.k8s.io/v1
kind: RuntimeClass
metadata:
  name: wasmedge
handler: wasmedge
```

## Reasoning
Maintaining the installation of different runtimes on different Linux and Kubernetes distributions is a complex task. KWasm will be redesigned to better address the requirement to install as many runtimes as possible, but at this stage only runwasi-based shims will be shipped because they have the simplest installation, while most wasm runtimes are available as runwasi shims.

Reply:
@0xE282B0 Thanks for your reminder. I update content related to KWasm. https://github.com/kubernetes/website/pull/43760/commits/a2cbd9575a0dd6738ae400bc19becc0a63085146